src/type_checker/fn_env.rs

```diff
@@ -110,14 +110,27 @@ impl TypedFnEnv {
         self.current_scope >= prefix_scope
     }
 
-    pub fn is_forbidden(&self, scope: usize) -> bool {
+    /// Since currently we don't support unrolling, the generic function calls are assumed to target a same instance.
+    /// This assumption requires a few type checking rules to forbid the cases that needs unrolling.
+    /// Forbid rules:
+    /// - Access to variables within the for loop scope.
+    /// - Access to mutable variables, except it is an array.
+    ///   Because once the array is declared, the size is fixed even if the array is mutable,
```

Comment:
ah, because generics use the array size, not the array content

Reply:
yes, exactly. generic use the array size.

Reply:
note that the type checking will throw error when you try to change the size of the array 
```
arr = truncate(arr); // throws error if the return of truncate() is a different type from arr
```